File: src/mcp/__tests__/validation.test.mjs
import { GkillApiError } from "../lib/errors.mjs";
import {
  assertObject,
  assertBoolean,
  assertNumber,
  assertInteger,
  assertTrimmedString,
  assertStringArray,
  assertIntegerArray,
  assertKnownKeys,
} from "../lib/validation.mjs";

// ---------------------------------------------------------------------------
// assertObject
// ---------------------------------------------------------------------------
describe("assertObject", () => {
  test("returns a plain object unchanged", () => {
    const obj = { a: 1 };
    expect(assertObject(obj, "field")).toBe(obj);
  });

  test("returns empty object", () => {
    const obj = {};
    expect(assertObject(obj, "field")).toBe(obj);
  });

  test("throws for null", () => {
    expect(() => assertObject(null, "field")).toThrow(GkillApiError);
  });

  test("throws for array", () => {
    expect(() => assertObject([1, 2], "field")).toThrow(GkillApiError);
  });

  test("throws for string", () => {
    expect(() => assertObject("hello", "field")).toThrow(GkillApiError);
  });

  test("throws for number", () => {
    expect(() => assertObject(42, "field")).toThrow(GkillApiError);
  });

  test("throws for undefined by default", () => {
    expect(() => assertObject(undefined, "field")).toThrow(GkillApiError);
  });

  test("returns undefined when allowUndefined is true", () => {
    expect(assertObject(undefined, "field", { allowUndefined: true })).toBeUndefined();
  });

  test("still throws for null even when allowUndefined is true", () => {
    expect(() => assertObject(null, "field", { allowUndefined: true })).toThrow(GkillApiError);
  });
});

// ---------------------------------------------------------------------------
// assertBoolean
// ---------------------------------------------------------------------------
describe("assertBoolean", () => {
  test("returns true", () => {
    expect(assertBoolean(true, "field")).toBe(true);
  });

  test("returns false", () => {
    expect(assertBoolean(false, "field")).toBe(false);
  });

  test("throws for number 0", () => {
    expect(() => assertBoolean(0, "field")).toThrow(GkillApiError);
  });

  test("throws for number 1", () => {
    expect(() => assertBoolean(1, "field")).toThrow(GkillApiError);
  });

  test("throws for string", () => {
    expect(() => assertBoolean("true", "field")).toThrow(GkillApiError);
  });

  test("throws for null", () => {
    expect(() => assertBoolean(null, "field")).toThrow(GkillApiError);
  });

  test("throws for undefined", () => {
    expect(() => assertBoolean(undefined, "field")).toThrow(GkillApiError);
  });
});

// ---------------------------------------------------------------------------
// assertNumber
// ---------------------------------------------------------------------------
describe("assertNumber", () => {
  test("returns valid number", () => {
    expect(assertNumber(3.14, "field")).toBe(3.14);
  });

  test("returns zero", () => {
    expect(assertNumber(0, "field")).toBe(0);
  });

  test("returns negative number", () => {
    expect(assertNumber(-5, "field")).toBe(-5);
  });

  test("throws for NaN", () => {
    expect(() => assertNumber(NaN, "field")).toThrow(GkillApiError);
  });

  test("throws for Infinity", () => {
    expect(() => assertNumber(Infinity, "field")).toThrow(GkillApiError);
  });

  test("throws for -Infinity", () => {
    expect(() => assertNumber(-Infinity, "field")).toThrow(GkillApiError);
  });

  test("throws for string", () => {
    expect(() => assertNumber("3.14", "field")).toThrow(GkillApiError);
  });

  test("respects min constraint", () => {
    expect(assertNumber(5, "field", { min: 5 })).toBe(5);
    expect(() => assertNumber(4, "field", { min: 5 })).toThrow(/greater than or equal to 5/);
  });

  test("respects max constraint", () => {
    expect(assertNumber(10, "field", { max: 10 })).toBe(10);
    expect(() => assertNumber(11, "field", { max: 10 })).toThrow(/less than or equal to 10/);
  });

  test("respects minExclusive constraint", () => {
    expect(assertNumber(0.001, "field", { minExclusive: 0 })).toBe(0.001);
    expect(() => assertNumber(0, "field", { minExclusive: 0 })).toThrow(/greater than 0/);
  });
});

// ---------------------------------------------------------------------------
// assertInteger
// ---------------------------------------------------------------------------
describe("assertInteger", () => {
  test("returns valid integer", () => {
    expect(assertInteger(42, "field")).toBe(42);
  });

  test("returns zero", () => {
    expect(assertInteger(0, "field")).toBe(0);
  });

  test("returns negative integer", () => {
    expect(assertInteger(-10, "field")).toBe(-10);
  });

  test("throws for float", () => {
    expect(() => assertInteger(3.14, "field")).toThrow(GkillApiError);
  });

  test("throws for string", () => {
    expect(() => assertInteger("42", "field")).toThrow(GkillApiError);
  });

  test("throws for NaN", () => {
    expect(() => assertInteger(NaN, "field")).toThrow(GkillApiError);
  });

  test("respects min constraint", () => {
    expect(assertInteger(0, "field", { min: 0 })).toBe(0);
    expect(() => assertInteger(-1, "field", { min: 0 })).toThrow(/greater than or equal to 0/);
  });

  test("respects max constraint", () => {
    expect(assertInteger(86399, "field", { max: 86399 })).toBe(86399);
    expect(() => assertInteger(86400, "field", { max: 86399 })).toThrow(/less than or equal to 86399/);
  });
});

// ---------------------------------------------------------------------------
// assertTrimmedString
// ---------------------------------------------------------------------------
describe("assertTrimmedString", () => {
  test("returns trimmed string", () => {
    expect(assertTrimmedString("hello", "field")).toBe("hello");
  });

  test("trims whitespace", () => {
    expect(assertTrimmedString("  hello  ", "field")).toBe("hello");
  });

  test("throws for empty string", () => {
    expect(() => assertTrimmedString("", "field")).toThrow(GkillApiError);
  });

  test("throws for whitespace-only string", () => {
    expect(() => assertTrimmedString("   ", "field")).toThrow(GkillApiError);
  });

  test("throws for number", () => {
    expect(() => assertTrimmedString(42, "field")).toThrow(GkillApiError);
  });

  test("throws for null", () => {
    expect(() => assertTrimmedString(null, "field")).toThrow(GkillApiError);
  });

  test("throws for boolean", () => {
    expect(() => assertTrimmedString(true, "field")).toThrow(GkillApiError);
  });
});

// ---------------------------------------------------------------------------
// assertStringArray
// ---------------------------------------------------------------------------
describe("assertStringArray", () => {
  test("returns array of trimmed strings", () => {
    expect(assertStringArray(["a", " b ", "c"], "field")).toEqual(["a", "b", "c"]);
  });

  test("returns empty array", () => {
    expect(assertStringArray([], "field")).toEqual([]);
  });

  test("throws for non-array", () => {
    expect(() => assertStringArray("not-array", "field")).toThrow(GkillApiError);
  });

  test("throws for null", () => {
    expect(() => assertStringArray(null, "field")).toThrow(GkillApiError);
  });

  test("throws if array contains non-string", () => {
    expect(() => assertStringArray(["ok", 42], "field")).toThrow(GkillApiError);
  });

  test("throws if array contains empty string", () => {
    expect(() => assertStringArray(["ok", ""], "field")).toThrow(GkillApiError);
  });

  test("error message includes index", () => {
    expect(() => assertStringArray(["ok", 42], "tags")).toThrow(/tags\[1\]/);
  });
});

// ---------------------------------------------------------------------------
// assertIntegerArray
// ---------------------------------------------------------------------------
describe("assertIntegerArray", () => {
  test("returns array of integers", () => {
    expect(assertIntegerArray([0, 3, 6], "field")).toEqual([0, 3, 6]);
  });

  test("returns empty array", () => {
    expect(assertIntegerArray([], "field")).toEqual([]);
  });

  test("throws for non-array", () => {
    expect(() => assertIntegerArray("not-array", "field")).toThrow(GkillApiError);
  });

  test("throws if element is float", () => {
    expect(() => assertIntegerArray([1, 2.5], "field")).toThrow(GkillApiError);
  });

  test("respects min/max constraints on elements", () => {
    expect(assertIntegerArray([0, 6], "field", { min: 0, max: 6 })).toEqual([0, 6]);
    expect(() => assertIntegerArray([0, 7], "field", { min: 0, max: 6 })).toThrow(/less than or equal to 6/);
    expect(() => assertIntegerArray([-1, 3], "field", { min: 0, max: 6 })).toThrow(/greater than or equal to 0/);
  });

  test("error message includes index", () => {
    expect(() => assertIntegerArray([0, "x"], "days")).toThrow(/days\[1\]/);
  });
});

// ---------------------------------------------------------------------------
// assertKnownKeys
// ---------------------------------------------------------------------------
describe("assertKnownKeys", () => {
  test("does not throw for known keys", () => {
    const allowed = new Set(["a", "b", "c"]);
    expect(() => assertKnownKeys({ a: 1, b: 2 }, allowed, "args")).not.toThrow();
  });

  test("does not throw for empty object", () => {
    const allowed = new Set(["a", "b"]);
    expect(() => assertKnownKeys({}, allowed, "args")).not.toThrow();
  });

  test("throws for unknown key", () => {
    const allowed = new Set(["a", "b"]);
    expect(() => assertKnownKeys({ a: 1, z: 2 }, allowed, "args")).toThrow(GkillApiError);
  });

  test("error includes unknown key name", () => {
    const allowed = new Set(["a"]);
    expect(() => assertKnownKeys({ unknown_key: 1 }, allowed, "args")).toThrow(/args\.unknown_key/);
  });

  test("error detail includes allowed keys", () => {
    const allowed = new Set(["b", "a"]);
    try {
      assertKnownKeys({ z: 1 }, allowed, "args");
      expect.unreachable("should have thrown");
    } catch (err) {
      expect(err.detail.allowed).toEqual(["a", "b"]);
    }
  });
});

File: src/server/gkill/mcp/validation_test.go
package mcp

// 旧 src/mcp/__tests__/validation.test.mjs の移植（describe → Test 関数、test → t.Run）。

import (
	"math"
	"testing"

	"github.com/mt3hr/gkill/src/server/gkill/mcp/jsonobj"
)

func TestAssertObject(t *testing.T) {
	t.Run("returns a plain object unchanged", func(t *testing.T) {
		o := obj("a", 1)
		got, err := AssertObject(o, "field")
		expectNoError(t, err)
		expectTrue(t, got == o, "must return the same object")
	})
	t.Run("returns empty object", func(t *testing.T) {
		o := obj()
		got, err := AssertObject(o, "field")
		expectNoError(t, err)
		expectTrue(t, got == o, "must return the same object")
	})
	t.Run("throws for null", func(t *testing.T) {
		_, err := AssertObject(nil, "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws for array", func(t *testing.T) {
		_, err := AssertObject(arr(1, 2), "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws for string", func(t *testing.T) {
		_, err := AssertObject("hello", "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws for number", func(t *testing.T) {
		_, err := AssertObject(42, "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws for undefined by default", func(t *testing.T) {
		_, err := AssertObject(jsonobj.Undefined, "field")
		expectGkillApiError(t, err)
	})
	t.Run("returns undefined when allowUndefined is true", func(t *testing.T) {
		got, err := AssertObjectAllowUndefined(jsonobj.Undefined, "field")
		expectNoError(t, err)
		expectTrue(t, got == nil, "must be nil for undefined")
	})
	t.Run("still throws for null even when allowUndefined is true", func(t *testing.T) {
		_, err := AssertObjectAllowUndefined(nil, "field")
		expectGkillApiError(t, err)
	})
}

func TestAssertBoolean(t *testing.T) {
	t.Run("returns true", func(t *testing.T) {
		got, err := AssertBoolean(true, "field")
		expectNoError(t, err)
		expectTrue(t, got, "true")
	})
	t.Run("returns false", func(t *testing.T) {
		got, err := AssertBoolean(false, "field")
		expectNoError(t, err)
		expectTrue(t, !got, "false")
	})
	t.Run("throws for number 0", func(t *testing.T) {
		_, err := AssertBoolean(0, "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws for number 1", func(t *testing.T) {
		_, err := AssertBoolean(1, "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws for string", func(t *testing.T) {
		_, err := AssertBoolean("true", "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws for null", func(t *testing.T) {
		_, err := AssertBoolean(nil, "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws for undefined", func(t *testing.T) {
		_, err := AssertBoolean(jsonobj.Undefined, "field")
		expectGkillApiError(t, err)
	})
}

func TestAssertNumber(t *testing.T) {
	t.Run("returns valid number", func(t *testing.T) {
		got, err := AssertNumber(3.14, "field", NumberRange{})
		expectNoError(t, err)
		expectTrue(t, got == 3.14, "got %v", got)
	})
	t.Run("returns zero", func(t *testing.T) {
		got, err := AssertNumber(0, "field", NumberRange{})
		expectNoError(t, err)
		expectTrue(t, got == 0, "got %v", got)
	})
	t.Run("returns negative number", func(t *testing.T) {
		got, err := AssertNumber(-5, "field", NumberRange{})
		expectNoError(t, err)
		expectTrue(t, got == -5, "got %v", got)
	})
	t.Run("throws for NaN", func(t *testing.T) {
		_, err := AssertNumber(math.NaN(), "field", NumberRange{})
		expectGkillApiError(t, err)
	})
	t.Run("throws for Infinity", func(t *testing.T) {
		_, err := AssertNumber(math.Inf(1), "field", NumberRange{})
		expectGkillApiError(t, err)
	})
	t.Run("throws for -Infinity", func(t *testing.T) {
		_, err := AssertNumber(math.Inf(-1), "field", NumberRange{})
		expectGkillApiError(t, err)
	})
	t.Run("throws for string", func(t *testing.T) {
		_, err := AssertNumber("3.14", "field", NumberRange{})
		expectGkillApiError(t, err)
	})
	t.Run("respects min constraint", func(t *testing.T) {
		got, err := AssertNumber(5, "field", NumberRange{Min: F(5)})
		expectNoError(t, err)
		expectTrue(t, got == 5, "got %v", got)
		_, err = AssertNumber(4, "field", NumberRange{Min: F(5)})
		expectErrorMatches(t, err, `greater than or equal to 5`)
	})
	t.Run("respects max constraint", func(t *testing.T) {
		got, err := AssertNumber(10, "field", NumberRange{Max: F(10)})
		expectNoError(t, err)
		expectTrue(t, got == 10, "got %v", got)
		_, err = AssertNumber(11, "field", NumberRange{Max: F(10)})
		expectErrorMatches(t, err, `less than or equal to 10`)
	})
	t.Run("respects minExclusive constraint", func(t *testing.T) {
		got, err := AssertNumber(0.001, "field", NumberRange{MinExclusive: F(0)})
		expectNoError(t, err)
		expectTrue(t, got == 0.001, "got %v", got)
		_, err = AssertNumber(0, "field", NumberRange{MinExclusive: F(0)})
		expectErrorMatches(t, err, `greater than 0`)
	})
}

func TestAssertInteger(t *testing.T) {
	t.Run("returns valid integer", func(t *testing.T) {
		got, err := AssertInteger(42, "field", IntRange{})
		expectNoError(t, err)
		expectTrue(t, got == 42, "got %v", got)
	})
	t.Run("returns zero", func(t *testing.T) {
		got, err := AssertInteger(0, "field", IntRange{})
		expectNoError(t, err)
		expectTrue(t, got == 0, "got %v", got)
	})
	t.Run("returns negative integer", func(t *testing.T) {
		got, err := AssertInteger(-10, "field", IntRange{})
		expectNoError(t, err)
		expectTrue(t, got == -10, "got %v", got)
	})
	t.Run("throws for float", func(t *testing.T) {
		_, err := AssertInteger(3.14, "field", IntRange{})
		expectGkillApiError(t, err)
	})
	t.Run("throws for string", func(t *testing.T) {
		_, err := AssertInteger("42", "field", IntRange{})
		expectGkillApiError(t, err)
	})
	t.Run("throws for NaN", func(t *testing.T) {
		_, err := AssertInteger(math.NaN(), "field", IntRange{})
		expectGkillApiError(t, err)
	})
	t.Run("respects min constraint", func(t *testing.T) {
		got, err := AssertInteger(0, "field", IntRange{Min: I(0)})
		expectNoError(t, err)
		expectTrue(t, got == 0, "got %v", got)
		_, err = AssertInteger(-1, "field", IntRange{Min: I(0)})
		expectErrorMatches(t, err, `greater than or equal to 0`)
	})
	t.Run("respects max constraint", func(t *testing.T) {
		got, err := AssertInteger(86399, "field", IntRange{Max: I(86399)})
		expectNoError(t, err)
		expectTrue(t, got == 86399, "got %v", got)
		_, err = AssertInteger(86400, "field", IntRange{Max: I(86399)})
		expectErrorMatches(t, err, `less than or equal to 86399`)
	})
}

func TestAssertTrimmedString(t *testing.T) {
	t.Run("returns trimmed string", func(t *testing.T) {
		got, err := AssertTrimmedString("hello", "field")
		expectNoError(t, err)
		expectTrue(t, got == "hello", "got %q", got)
	})
	t.Run("trims whitespace", func(t *testing.T) {
		got, err := AssertTrimmedString("  hello  ", "field")
		expectNoError(t, err)
		expectTrue(t, got == "hello", "got %q", got)
	})
	t.Run("throws for empty string", func(t *testing.T) {
		_, err := AssertTrimmedString("", "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws for whitespace-only string", func(t *testing.T) {
		_, err := AssertTrimmedString("   ", "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws for number", func(t *testing.T) {
		_, err := AssertTrimmedString(42, "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws for null", func(t *testing.T) {
		_, err := AssertTrimmedString(nil, "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws for boolean", func(t *testing.T) {
		_, err := AssertTrimmedString(true, "field")
		expectGkillApiError(t, err)
	})
}

func TestAssertStringArray(t *testing.T) {
	t.Run("returns array of trimmed strings", func(t *testing.T) {
		got, err := AssertStringArray(strs("a", " b ", "c"), "field")
		expectNoError(t, err)
		expectEqual(t, got, strs("a", "b", "c"))
	})
	t.Run("returns empty array", func(t *testing.T) {
		got, err := AssertStringArray(arr(), "field")
		expectNoError(t, err)
		expectEqual(t, got, arr())
	})
	t.Run("throws for non-array", func(t *testing.T) {
		_, err := AssertStringArray("not-array", "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws for null", func(t *testing.T) {
		_, err := AssertStringArray(nil, "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws if array contains non-string", func(t *testing.T) {
		_, err := AssertStringArray(arr("ok", 42), "field")
		expectGkillApiError(t, err)
	})
	t.Run("throws if array contains empty string", func(t *testing.T) {
		_, err := AssertStringArray(arr("ok", ""), "field")
		expectGkillApiError(t, err)
	})
	t.Run("error message includes index", func(t *testing.T) {
		_, err := AssertStringArray(arr("ok", 42), "tags")
		expectErrorMatches(t, err, `tags\[1\]`)
	})
}

func TestAssertIntegerArray(t *testing.T) {
	t.Run("returns array of integers", func(t *testing.T) {
		got, err := AssertIntegerArray(arr(0, 3, 6), "field", IntRange{})
		expectNoError(t, err)
		expectEqual(t, got, arr(0, 3, 6))
	})
	t.Run("returns empty array", func(t *testing.T) {
		got, err := AssertIntegerArray(arr(), "field", IntRange{})
		expectNoError(t, err)
		expectEqual(t, got, arr())
	})
	t.Run("throws for non-array", func(t *testing.T) {
		_, err := AssertIntegerArray("not-array", "field", IntRange{})
		expectGkillApiError(t, err)
	})
	t.Run("throws if element is float", func(t *testing.T) {
		_, err := AssertIntegerArray(arr(1, 2.5), "field", IntRange{})
		expectGkillApiError(t, err)
	})
	t.Run("respects min/max constraints on elements", func(t *testing.T) {
		got, err := AssertIntegerArray(arr(0, 6), "field", IntRange{Min: I(0), Max: I(6)})
		expectNoError(t, err)
		expectEqual(t, got, arr(0, 6))
		_, err = AssertIntegerArray(arr(0, 7), "field", IntRange{Min: I(0), Max: I(6)})
		expectErrorMatches(t, err, `less than or equal to 6`)
		_, err = AssertIntegerArray(arr(-1, 3), "field", IntRange{Min: I(0), Max: I(6)})
		expectErrorMatches(t, err, `greater than or equal to 0`)
	})
	t.Run("error message includes index", func(t *testing.T) {
		_, err := AssertIntegerArray(arr(0, "x"), "days", IntRange{})
		expectErrorMatches(t, err, `days\[1\]`)
	})
}

func TestAssertKnownKeys(t *testing.T) {
	t.Run("does not throw for known keys", func(t *testing.T) {
		allowed := NewStringSet("a", "b", "c")
		expectNoError(t, AssertKnownKeys(obj("a", 1, "b", 2), allowed, "args", nil))
	})
	t.Run("does not throw for empty object", func(t *testing.T) {
		allowed := NewStringSet("a", "b")
		expectNoError(t, AssertKnownKeys(obj(), allowed, "args", nil))
	})
	t.Run("throws for unknown key", func(t *testing.T) {
		allowed := NewStringSet("a", "b")
		expectGkillApiError(t, AssertKnownKeys(obj("a", 1, "z", 2), allowed, "args", nil))
	})
	t.Run("error includes unknown key name", func(t *testing.T) {
		allowed := NewStringSet("a")
		expectErrorMatches(t, AssertKnownKeys(obj("unknown_key", 1), allowed, "args", nil), `args\.unknown_key`)
	})
	t.Run("error detail includes allowed keys", func(t *testing.T) {
		allowed := NewStringSet("b", "a")
		err := AssertKnownKeys(obj("z", 1), allowed, "args", nil)
		apiErr := expectGkillApiError(t, err)
		expectEqual(t, apiErr.Detail.Value("allowed"), strs("a", "b"))
	})
}

func TestAssertKnownKeysDefaultFieldName(t *testing.T) {
	t.Run("unknown keys are reported under arguments when no field is given", func(t *testing.T) {
		// 第3引数を渡さない呼び出しが write 側に20箇所あり、
		// エラーが `Invalid argument 'undefined.contnet'` になっていた
		err := AssertKnownKeys(obj("content", "x", "contnet", "y"), NewStringSet("content"), "", nil)
		apiErr := expectGkillApiError(t, err)
		expectEqual(t, apiErr.Detail.Value("field"), "arguments.contnet")
		mustNotContain(t, apiErr.Message, "undefined")
	})
}

func TestAssertIntegerSafeIntegerRange(t *testing.T) {
	// assertInteger — 安全整数の範囲 (2巡目の指摘 P-37)
	t.Run("rejects integers beyond the safe range instead of storing a value that cannot round-trip", func(t *testing.T) {
		// 2^53。JSON を往復するだけで別の値になるので、保存できたように見えて読み戻すと違う
		_, err := AssertInteger(float64(9007199254740992), "amount", IntRange{})
		expectErrorMatches(t, err, `safe integer range`)
		_, err = AssertInteger(float64(-9007199254740992), "amount", IntRange{})
		expectErrorMatches(t, err, `safe integer range`)
	})
	t.Run("still accepts the largest value that does round-trip", func(t *testing.T) {
		got, err := AssertInteger(jsonobj.MaxSafeInteger, "amount", IntRange{})
		expectNoError(t, err)
		expectTrue(t, float64(got) == jsonobj.MaxSafeInteger, "got %v", got)
		got, err = AssertInteger(-jsonobj.MaxSafeInteger, "amount", IntRange{})
		expectNoError(t, err)
		expectTrue(t, float64(got) == -jsonobj.MaxSafeInteger, "got %v", got)
	})
	t.Run("the safe-range error is a GkillApiError like every other validation error", func(t *testing.T) {
		_, err := AssertInteger(math.Pow(2, 53), "amount", IntRange{})
		expectGkillApiError(t, err)
	})
}
